src/Plugin/EndpointBase.php

```diff
@@ -0,0 +1,301 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->configuration += $this->defaultConfiguration();
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getScheme() {
+    return $this->scheme;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getAuthority() {
+    if (empty($this->host)) {
+      return '';
+    }
+    $authority = $this->host;
+    if (!empty($this->userInfo)) {
+      $authority = $this->userInfo . '@' . $authority;
+    }
+    if (!empty($this->port)) {
+      $authority .= ':' . $this->port;
+    }
+    return $authority;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getUserInfo() {
+    return $this->userInfo;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getHost() {
+    return $this->host;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getPort() {
+    return $this->port;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getPath() {
+    return $this->path == null ? '' : $this->path;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getQuery() {
+    return $this->query;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getFragment() {
+    return $this->fragment;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withScheme($scheme) {
+    $new = clone $this;
+    $new->scheme = $scheme;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withUserInfo($user, $password = NULL) {
+    $info = $user;
+    if ($password) {
+      $info .= ':' . $password;
+    }
+
+    $new = clone $this;
+    $new->userInfo = $info;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withHost($host) {
+    $new = clone $this;
+    $new->host = $host;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withPort($port) {
+    $new = clone $this;
+    $new->port = $port;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withPath($path) {
+    $new = clone $this;
+    $new->path = $path;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withQuery($query) {
+    $new = clone $this;
+    $new->query = $query;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function withFragment($fragment) {
+    $new = clone $this;
+    $new->fragment = $fragment;
+    return $new;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function __toString() {
+    $uri = '';
+    if (!empty($this->scheme)) {
+      $uri .= $this->scheme . '://';
+    }
+    if (!empty($this->getAuthority())) {
+      $uri .= $this->getAuthority();
+    }
+    if ($this->getPath() != null) {
+      // Add a leading slash if necessary.
+      if ($uri && substr($this->getPath(), 0, 1) !== '/') {
+        $uri .= '/';
+      }
+      $uri .= $this->getPath();
+    }
+    if ($this->query != null) {
+      $uri .= '?' . $this->query;
+    }
+    if ($this->fragment != null) {
+      $uri .= '#' . $this->fragment;
+    }
+    return $uri;
```

Comment:
So a simple Unit test would be enough I think, e.g. just instantiate the bare classes with some bogus parameters and test the output of all the `with*` methods and `__toString()` etc.